File: Session12/C++/TopViewOfBinaryTree.cpp
https://practice.geeksforgeeks.org/problems/top-view-of-binary-tree/1

class Solution
{
    public:
    //Function to return a list of nodes visible from the top view 
    //from left to right in Binary Tree.
    vector<int> topView(Node *root)
    {
        //Your code here
        queue<pair<Node*, int>> q;
        q.push({root, 0});
        vector<int> res;
        map<int, int> topNodeMap;
        
        while(!q.empty())
        {
            Node* node = q.front().first;
            int hd = q.front().second;
            q.pop();
            
            if(!topNodeMap[hd]) {
                topNodeMap[hd] = node->data;
            }
            
            if(node->left) {
                q.push({node->left, hd-1});
            }
            if(node->right) {
                q.push({node->right, hd+1});
            }
        }
        
        for(auto it:topNodeMap) {
            res.push_back(it.second);
        }
        return res;
    }
};

File: Session12/Java/TopViewOfBinaryTree.java
https://practice.geeksforgeeks.org/problems/top-view-of-binary-tree/1

class Solution
{
    static class Pair
    {
        Node node;
        int hd;
        Pair(Node node, int hd){
            this.node = node;
            this.hd = hd;
        }
    }
    //Function to return a list of nodes visible from the top view 
    //from left to right in Binary Tree.
    static ArrayList<Integer> topView(Node root)
    {
        // add your code
        ArrayList<Integer> ans = new ArrayList<>(); 
        if(root == null) 
            return ans;
        Map<Integer, Integer> topNodeMap = new TreeMap<>();
        Queue<Pair> q = new LinkedList<Pair>();
        q.add(new Pair(root, 0)); 
        
        while(!q.isEmpty()) {
            Pair pair = q.remove();
            int hd = pair.hd; 
            Node node = pair.node; 
            if(topNodeMap.get(hd) == null){
                topNodeMap.put(hd, node.data);    
            } 
            
            if(node.left != null) {
                q.add(new Pair(node.left, hd - 1)); 
            }
            if(node.right != null) {
                q.add(new Pair(node.right, hd + 1)); 
            }
        }
    
        for (Map.Entry<Integer,Integer> entry : topNodeMap.entrySet()) {
            ans.add(entry.getValue()); 
        }
        return ans; 
    }
}